# ccdbgUpdater
ccdbgUpdater is a modular tool for updating a colored and compacted de Bruijn Graph. 

### Augmenting a compacted de Bruijn Graph by a new genome or a set of sequences


## Requirements

ccdbgUpdater is developed in C++11
* C++11 compiler:
    * [GCC](https://gcc.gnu.org/) >= 4.8.5

It relies on kmtricks to find the set of absent k-mers which are in the genome to be added but not in the graph
## Installation
Installing kmtricks: First Install kmtricks
* From [Bioconda](https://bioconda.github.io):

  ```
  conda install -c conda-forge -c tlemane kmtricks
  ```
Installing ccdbupdater: Currently we have only the source version
* From source

  ```
  git clone https://github.com/khodor14/ccdbgUpdater.git
  cd ccdbgUpdater/src
  make
  ```

## Binary usage:

```
./main
```

displays the command line interface:
```
Usage:
./main --input_graph <value> --input_genome <value> --k_mer_size <value>
OR
./main --input_graph <value> --k_mer_file <value> --k_mer_size <value>



Options:
	-h[--help]	 prints this help message
	--input_graph	 the path to the pangenome graph in gfa or fasta format
	--input_genome	 the path to the genome used to augment the input graph
	--k_mer_size[-k] the size of the k-mer.
			 It must be the same value used when constructing the input graph
	--k_mer_file	 the file of absent k-mers from the graph if already computed
	--test		 used in testing mode to compare
			 the augmented graph generated by this algo
			 to an already augmented graph
			 if this is set to true then the augmented graph must be given as input
	--augmented_graph the path to an already augmented graph in fasta or gfa
			 it must be given if the argument --test is set to true
	--output_file_name[-o] the name of the output file
	-v verbosity
	--output_kmers output kmers to temporary file

```

### Examples

- 

  1. **Update a compacted de bruijn graph from a kmer file**
     ```
     ./main --input_graph graph.gfa --k_mer_file kmers.txt -k 31 -o output_prefix
     ```
     The compacted de Bruijn graph *graph.gfa* is updated  from the 31-mers (`-k 31`) of files *kmers.txt*. The updated graph is stored in *output_prefix.fa*

  2. **Update a compacted de Bruijn graph from a reference genome file**
     ```
     ./main --input_graph graph.gfa --k_mer_file kmers.txt -k 31 -o output_prefix
     ```
     The compacted de Bruijn graph is built (`build`) with 4 threads (`-t 4`) from the 31-mers (`-k 31`) of file *C.fasta* (`-r C.fasta`). By using parameter `-r`, file *C.fasta* is NOT filtered: all 31-mers occurring in *C* are used during the construction. The graph is written to file *C_graph.gfa* (`-o C_graph`).



### Algorithm
- 

  1. **Using kmtricks**
    Three commands of kmtricks are used to find the absent kmers
     ```
     kmtricks pipeline --file fof.txt --run-dir graph_dir  --kmer-size 31 --hard-min 1 --mode kmer:pa:bin --nb-partitions 100 --cpr --threads 32
     ```
     ```
     kmtricks filter --in-matrix graph_dir --key query.txt --output filter --hard-min 1 --out-types k --cpr-in --cpr-out -t 32
     ```
    ```
     kmtricks aggregate --run-dir filter --count A1:kmer --format text --cpr-in --output kmers.txt
    ```
    For more details on the documentation of kmtricks visit https://github.com/tlemane/kmtricks/wiki

  2. **Indexing the input Graph**
     We store (k-1)-mers of the unitig of the graph as (k-1)-mer->Vector[(unitig id,position,orientation)]

  3. **Constructing the unitigs from absent k-mers with respect to the graph**
     We go over the absent k-mers, we try to extend each by its in neighbor k-mer from left and right. This process is implemented by following
     the node centric definition of a de bruijn graph and by ensuring that the extended unitig does not branch to graph. 
     If it branches or it breaks the definition, we break the extension. We use a flag to tell if a k-mer was used before or not
    
  4. **Indexing the contructed unitigs**
    Once the unitigs are constructed, we index them by their (k-1)-mer prefixes and suffixes
  5. **Updating the input graph**
    Finally we go over the index built in 4, we determine the decision (split,merge) by evaluating the position of the prefix and its multiplicity 
    both in the unitigs of the graph and the constructed unitigs

   6. **Updating the index of the graph**
    We update the index of the graph on the fly.

## Citation
This work is on progress, but it relies on kmtricks. If you use it cite kmtricks:
```
@article {10.1093/bioadv/vbac029,
    author = {Lemane, TÃ©o and Medvedev, Paul and Chikhi, Rayan and Peterlongo, Pierre},
    title = "{kmtricks: efficient and flexible construction of Bloom filters for large sequencing data collections}",
    journal = {Bioinformatics Advances},
    volume = {2},
    number = {1},
    year = {2022},
    month = {04},
}
```

## Contact

For any question, feedback or problem, please feel free to file an issue on this GitHub repository and we will get back to you as soon as possible.
